drivers/sdhc/sdhc_renesas_ra.c

```diff
@@ -0,0 +1,718 @@
+	enum sdhc_timing_mode timing;
+	enum sdhc_power power_mode;
+	struct k_sem thread_lock;
+	uint8_t status;
+	struct sdhc_host_props props;
+	/* Transfer DTC */
+	struct st_transfer_instance transfer;
+	struct st_dtc_instance_ctrl transfer_ctrl;
+	struct st_transfer_info transfer_info;
+	struct st_transfer_cfg transfer_cfg;
+	struct st_dtc_extended_cfg transfer_cfg_extend;
+};
+
+void sdhimmc_accs_isr(void);
+void sdhimmc_card_isr(void);
+void sdhimmc_dma_req_isr(void);
+
+static void ra_sdmmc_accs_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_accs_isr();
+}
+
+static void ra_sdmmc_card_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_card_isr();
+}
+
+static void ra_sdmmc_dma_req_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_dma_req_isr();
+}
+
+static volatile struct sdmmc_ra_event sdmmc_event = {
+	.transfer_completed = false,
+	.transfer_error = false,
+};
+
+void r_sdhi_callback(sdmmc_callback_args_t *p_args)
+{
+	if (p_args->event == SDMMC_EVENT_TRANSFER_COMPLETE) {
+		sdmmc_event.transfer_completed = true;
+		sdmmc_event.transfer_error = false;
+	} else if (p_args->event == SDMMC_EVENT_TRANSFER_ERROR) {
+		sdmmc_event.transfer_error = true;
+		sdmmc_event.transfer_completed = false;
+	}
+}
+
+static int sdhc_ra_get_card_present(const struct device *dev)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int ret;
+	sdmmc_status_t status;
+
+	/* SDMMC_CARD_DETECT_CD must be configured as true to check here */
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.card_inserted);
+}
+
+static int sdhc_ra_card_busy(const struct device *dev)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int ret;
+	sdmmc_status_t status;
+
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.transfer_in_progress);
+}
+
+static int sdhc_ra_send_cmd(struct sdhc_ra_priv *priv, struct sdmmc_ra_command *ra_cmd, int retries)
+{
+	int fsp_err = 0;
+
+	while (retries > 0) {
+		fsp_err = r_sdhi_command_send(&priv->sdmmc_ctrl, ra_cmd->opcode, ra_cmd->arg);
+		if (fsp_err != 0) {
+			retries--; /* error, retry */
+		} else {
+			break;
+		}
+	}
+	return err_fsp2zep(fsp_err);
+}
+
+/*
+ * Send CMD or CMD/DATA via SDHC
+ */
+static int sdhc_ra_request(const struct device *dev, struct sdhc_command *cmd,
+			   struct sdhc_data *data)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	int retries = (int)(cmd->retries + 1); /* first try plus retries */
+	fsp_err_t fsp_err = 0;
+	uint32_t timeout_cfg = 0;
+	int ret = 0;
+	sdmmc_priv_csd_reg_t p_csd_reg;
+
+	struct sdmmc_ra_command ra_cmd = {
+		.opcode = cmd->opcode,
+		.arg = cmd->arg,
+	};
+
+	if (data) {
+		ra_cmd.data = (uint8_t *)data->data;
+		ra_cmd.sector_start = data->block_addr;
+		ra_cmd.sector_count = data->blocks;
+		timeout_cfg = data->timeout_ms;
```

Comment:
This isn't used, but needs to be. You need to respect the timeout configuration in some way- typically this done by waiting for an command completion signal (either via polling or interrupt) until the timeout expires. You should also be waiting for the command timeout, *then* the data timeout. The data timeout does not take precedence over the command timeout when present.

Reply:
updated to wait the command timeout by `sdhi_command_send_wait`.